### PR TITLE
Quadratic elements for peripheral ring

### LINKED ISSUE #28169: Enable Quadratic Elements in PeripheralRingMeshGenerator
### Motivation

Original Reactor Module tools are only capable of creating meshes with linear elements, which limits its applications. 

Efforts have been made to enable quadratic element capabilities. This enhancement follows up on #26986 #28164

### Design

Provide quadratic element (QUAD9) as an option in `PeripheralRingMeshGenerator`.

### Impact

Quadratic element capability would benefit users with related needs and thus enhance the Reactor Module as well as MOOSE's intrinsic meshing capabilities.

### PR DESCRIPTION
**Pull request by miaoyinb**
closes #28169

### CODE REVIEW

**Comment by miaoyinb** [comment]
@GiudGiud FYI. Currently it can only do QUAD4 and QUAD9, I can add QUAD8 if needed.

**Comment by miaoyinb** [comment]
@eshemon FYI

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:85]:
```suggestion
    requirement ='The system shall be able to add a peripheral ring consisting of quadratic elements with boundary layers to the input mesh and preserve volumes despite polygonization.'
```

File: framework/include/utils/FillBetweenPointVectorsTools.h
Comment (GiudGiud) [inline framework/include/utils/FillBetweenPointVectorsTools.h:299]:
```suggestion
 * @param node_assm the assembly of the node id pairs that correspond to vertices of the sides on the closed loop
```

File: framework/include/utils/FillBetweenPointVectorsTools.h
Comment (GiudGiud) [inline framework/include/utils/FillBetweenPointVectorsTools.h:303]:
```suggestion
 * azimuthal angles changing monotonically
```

File: framework/include/utils/FillBetweenPointVectorsTools.h
Comment (GiudGiud) [reply framework/include/utils/FillBetweenPointVectorsTools.h:299]: if it s correct

**Review by GiudGiud — COMMENTED** [review]
will look at the code latr

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C
Comment (pbehne) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:180]: Please add a `RunException` test for this.

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C
Comment (pbehne) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:184]: Please add a `RunException` test for this.

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C
Comment (pbehne) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:506]: Why the tolerance change?

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i
Comment (pbehne) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:12]: Why `QUAD8` instead of `QUAD9`? Just curious.

**Review by pbehne — CHANGES_REQUESTED** [review]
Please put a short description of what this PR does in the header of the PR, even if it is just copying and pasting the issue.

While reviewing, I cleaned up some of the changed files (not really related to your PR). I'd like to get these changes in with this PR. May I push to your branch (you may have to add me as a collaborator to your fork of MOOSE). If not, I can also make a PR with my changes to your fork (after you add me as a collaborator).

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i
Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:12]: this PR adds quad9 for the periphery. for the mesh inside, quad8 will use less memory and is just as compatible with the boundary

**Comment by miaoyinb** [comment]
> Please put a short description of what this PR does in the header of the PR, even if it is just copying and pasting the issue.
> 
> While reviewing, I cleaned up some of the changed files (not really related to your PR). I'd like to get these changes in with this PR. May I push to your branch (you may have to add me as a collaborator to your fork of MOOSE). If not, I can also make a PR with my changes to your fork (after you add me as a collaborator).

I think the PR template says it does not need any description if the issue # is linked in the header. Anyway I will add you as a collaborator, please do a PR there so that I could take a look. Thanks.

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:506]: The previous `PeripheralRingMeshGenerator::innerBdryLayerNodesDefiner()` has a bug. It would have issue in some cases when the inner boundary layer is too thick. This, along with the following additions, is part of the fix.

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:180]: Does current MOOSE have an easy way to load/create a third-order 2D mesh to trigger this error?

**Comment by pbehne** [comment]
> > Please put a short description of what this PR does in the header of the PR, even if it is just copying and pasting the issue.
> > While reviewing, I cleaned up some of the changed files (not really related to your PR). I'd like to get these changes in with this PR. May I push to your branch (you may have to add me as a collaborator to your fork of MOOSE). If not, I can also make a PR with my changes to your fork (after you add me as a collaborator).
> 
> I think the PR template says it does not need any description if the issue # is linked in the header. Anyway I will add you as a collaborator, please do a PR there so that I could take a look. Thanks.

Thanks for pointing out the template instructions. Since you have an issue number, could you go ahead and delete the empty headers (Reason, Design, Impact)?

PR is up here: https://github.com/miaoyinb/moose/pull/5

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C
Comment (pbehne) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:180]: Tagging @roystgnr on this one since I'm not sure. He will likely know off the top of his head.

Comment (pbehne) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:180]: I see that libmesh has a 1D `Edge4` element. I'll see if I can figure out what kind of element uses this.

Comment (pbehne) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:180]: As far as I can tell from looking through libmesh source, `Edge4` is not used as a side in any element. I'll let Roy have the final word, though.

Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:181]: Right, I do not see any complete third-order 2D elements there.

Comment (roystgnr) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:181]: We currently don't have any 3rd-order elements outside of 1D, so no way to add test coverage for this.

We might be adding some in the next year, so it's good that we're catching that possibility here, but not soon enough that I expect I'll remember this PR when we're finally able to add a test for it.

**Review by pbehne — APPROVED** [review]
Thanks for adding this feature! I'm ready to merge if @GiudGiud is.

**Comment by miaoyinb** [comment]
@GiudGiud Thank you for reviewing/merging my other PR. Just a friendly reminder that this PR is still pending.

**Comment by GiudGiud** [comment]
sorry for the delay. I am setting up a new machine the other one died by IT

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: If I  comment these two out to make the mesh linear, I get a different volumes for the external ring

**Review by GiudGiud — COMMENTED** [review]
there's something I dont understand with the test

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i
Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: I will look into this, thanks

Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: Interesting, paraview says the areas are the same...(706.858 total which is the same as the linear elements values reported by MOOSE)

Maybe we need to take a look at the source code that calculates these printed volume values...

Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]:
```
[Mesh]
  [dir1]
    type = ElementGenerator
    nodal_positions = '0 0 0
                       1 0 0
                       1 1 0
                       0 1 0
                       0.5 -0.5 0
                       1 0.5 0
                       0.5 1 0
                       0 0.5 0
                       0.5 0.5 0'
    element_connectivity = '0 1 2 3 4 5 6 7 8'
    elem_type = 'QUAD9'
  []
[]
```
The element area is 1.25 but MOOSE says 1.333 (or quadratic element area has a different definition?)

Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: @roystgnr thoughts?

Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: I think when you say 1.25 you have a triangle between the three slopes at the bottom
shouldnt it be quadratic?

Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: I just tried to give an extreme case. There might be some rules that I am not aware of. Let's see if @roystgnr has any insights.

Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: > I think when you say 1.25 you have a triangle between the three nodes at the bottom of the element shouldnt it be quadratic?

Ah, so the edge should be parabolic for QUAD9 element. In that case the radius correction function needs to be modified.

Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: I think so.
Irrelevant here but I think it might be the same for quad8?

Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: fixed. sorry for the batch regolding...

File: modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/gold/single_hex_quadratic_in.e
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/gold/single_hex_quadratic_in.e:1]: the nodes should not have moved for those? the modifying of the periphery was sqauree right?

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/gold/ring_boundary_layers_quadratic_in.e
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/gold/ring_boundary_layers_quadratic_in.e:1]: let's remove that gold file from commit #1 to avoid the regold

File: modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/gold/single_hex_quadratic_in.e
Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/gold/single_hex_quadratic_in.e:1]: true true

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: @roystgnr for a quad9, does the position of a middle node influence the shape of the outer boundary?

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1197]: MathUtils
```suggestion
          4.0 / 3.0 * Utility::pow<3>(std::sin((azi_interval_list[i].first + azi_interval_list[i].second) / 2)) /
```

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1199]: why both for this case?

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C
Comment (roystgnr) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: It does not.  Not for Lagrange mappings, not for Rational Bezier Bernstein mappings.

There are mappings with less locality than the ones we use, but we'll probably never support any directly, only indirectly the way we support IsoGeometric Analysis NURBS mappings - via extraction operators that give more locally-defined elements to integrate on.

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: thanks

how exact is the volume computation for second order elements in libmesh? I still see tiny diffs (1e-5) with the correction we have here

Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: I think that summation of small $\sin^3(\theta)$ compromises the errors...

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: could be

are you summing two separate quadratic bands, separated at the mid point of the edge?
I think the whole edge is covered by a single quadratic band?

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: > But are you diffing against the exact biquadratic volume, or against a perfect circle? Biquadratics (in Lagrange; I'm working on RBB mesh generator tools) aren't enough to exactly mesh a circle; no polynomial mapping is. The difference between a unit circle and its biquadratic approximation is probably more like 1e-5.

I am diffing against the corrected 1st order mesh, which should match the perfect circle volume exactly already.
While the bi-quadratics do not exactly mesh the circle, we are correcting them here so that the volume is the exact same. This is what this routine is intended for. So yes it's only correcting a very small error already

Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: > could be
> 
> are you summing two separate quadratic bands, separated at the mid point of the edge? I think the whole edge is covered by a single quadratic band?

Yes, it's a single band determined by the three nodes.

Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: If we really want to eliminate that 1e-6 relative error, one possible way would be to make a dummy TRI6 mesh with the same external nodes and use its area as the reference to derive a correction factor.

File: framework/src/utils/FillBetweenPointVectorsTools.C
Comment (pbehne) [reply framework/src/utils/FillBetweenPointVectorsTools.C:470]: Here is the compiler error when I don't cast to long:
`In file included from /Users/behnpa/projects/moose_quadratic_ring/framework/build/unity_src/utils_Unity.C:20:
/Users/behnpa/projects/moose_quadratic_ring/framework/src/utils/FillBetweenPointVectorsTools.C:470:27: error: no matching function for call to 'make_range'
      for (const auto k : make_range(1, it_dist))`
I tried using auto instead of long, and got:
`/Users/behnpa/projects/moose_quadratic_ring/framework/src/utils/FillBetweenPointVectorsTools.C:470:38: error: 'auto' not allowed here
      for (const auto k : make_range(auto(1), it_dist))
`

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i
Comment (roystgnr) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/ring_quadratic.i:11]: > Ah, so the edge should be parabolic for QUAD9 element

This is correct.  In the `ElementGenerator` case in particular, 4/3 is the correct area.

File: framework/src/utils/FillBetweenPointVectorsTools.C
Comment (pbehne) [reply framework/src/utils/FillBetweenPointVectorsTools.C:470]: For loops that start from 1, should we just use the other style of for loop?

Comment (GiudGiud) [reply framework/src/utils/FillBetweenPointVectorsTools.C:470]: unless you find a way to get the right data type for the 1 with various compiling options I would say yes

Comment (miaoyinb) [reply framework/src/utils/FillBetweenPointVectorsTools.C:470]: could be overkill but this would work...

`make_range(castIndexForRange(foo,1),foo)`

```
template <typename T, typename P>
const T
castIndexForRange(const T &, //type_input
                  const P & value_input)
{
  return (T)(value_input);
}
```

Comment (roystgnr) [reply framework/src/utils/FillBetweenPointVectorsTools.C:470]: `make_range()` is just some syntactic sugar.  It's just to make it easier to avoid the typical C-style mistake of `for (foo i=begin(); i != expensive_call_that_shouldnt_be_here(); ++i)`, not because there's anything else we could fix with range-for vs C-style-for loops.  (though I'll bet we have a lot of both kinds of loops that should be `std::foreach`, and I'm thinking of wrapping that to play with different `std::execution::` options...)

`castIndexForRange()` is syntactic Carolina Reaper powder.  Maybe it's an acquired taste?

I think C-style is fine; otherwise `1l` is more terse than `long(1)`

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C
Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: It s more that if we expect the error to be 1e-6 already with quad-elements (as Roy mentioned), then why do we still see that after correction

I m yet to run this again and make up my mind on this

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: it might just be that the error is a larger than expected pre-correction

Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1165]: I will also run more cases to get a clearer idea what is going on.

**Comment by GiudGiud** [comment]
saw your changes, I ll take a look tomorrow

**Comment by GiudGiud** [comment]
It looks like the background size in the square lattice, when deformation of non-circular regions is set to false, is different between quadratic and linear elements. It affects the regolded files for patterned_cartesian_mg (pattern_sq_) and I think that's it.

Which volume is most important to preserve in that case?

**Comment by GiudGiud** [comment]
Otherwise we are good on all fronts here

**Comment by miaoyinb** [comment]
> It looks like the background size in the square lattice, when deformation of non-circular regions is set to false, is different between quadratic and linear elements. It affects the regolded files for patterned_cartesian_mg (pattern_sq_) and I think that's it.
> 
> Which volume is most important to preserve in that case?

If the non-circular region deformation is activated, the background region is deformed to allow small hexagon/square size that cuts the original patterned unit cells. In that case, we do not care about the volume preserve as they are deformed anyway.

But if the deformation is not activated, the midpoint should be colinear with the vertices and the quandratic elements and linear elements should have the same sizes(?)

**Comment by GiudGiud** [comment]
I think the square size should be the same for both linear and quadractic though? We lose the volume conservation of the background but we impose a square size right?

the diff is in the case with the deformation

**Comment by miaoyinb** [comment]
> I think the square size should be the same for both linear and quadractic though? We lose the volume conservation of the background but we impose a square size right?
> 
> the diff is in the case with the deformation

Right.

**Comment by GiudGiud** [comment]
ok so it's not tied to this PR but seems there's a problem in the PatternedCartesian w/ deformation & second order.
pls let me know if you want to address it here with the regold or if we should create an issue

**Comment by miaoyinb** [comment]
Ah, I figured out the source of the issue. Let's address it here.

**Comment by miaoyinb** [comment]
That should be fixed now.

**Comment by GiudGiud** [comment]
this didnt fix it. If you zoom in on the corner you can see that pattern_sq.e and pattern_sq_quad8.e for the patterned_cartesian_mg dont line up at the pin background / duct background interface

**Comment by miaoyinb** [comment]
> this didnt fix it. If you zoom in on the corner you can see that pattern_sq.e and pattern_sq_quad8.e for the patterned_cartesian_mg dont line up at the pin background / duct background interface

Let me check it again. It seems the hex and cartesian have different issues...

**Comment by miaoyinb** [comment]
The boundary of the pin background and assembly background after deformation is okay to be different between linear and quadratic elements. The size of the outer boundary of the assembly (i.e., inner size of the innermost duct) is important and must be consistent.

**Review by GiudGiud — APPROVED** [review]
thanks for the contribution!

**Pull request merged**
